packages/venia-ui/lib/components/ProductFullDetail/customAttributes.js

```diff
@@ -0,0 +1,251 @@
+import React, { useMemo } from 'react';
+import { array, shape, string } from 'prop-types';
+import { FormattedDate, FormattedMessage, FormattedTime } from 'react-intl';
+
+import { BrowserPersistence } from '@magento/peregrine/lib/util';
+import { useStyle } from '@magento/venia-ui/lib/classify';
+import Price from '@magento/venia-ui/lib/components/Price';
+import RichContent from '@magento/venia-ui/lib/components/RichContent';
+
+import defaultClasses from './customAttributes.module.css';
+
+const storage = new BrowserPersistence();
+
+const CustomAttributes = props => {
+    const { customAttributes } = props;
+    const classes = useStyle(defaultClasses, props.classes);
+
+    // TODO: Get complete price data from GraphQl including currency
+    const storeCurrency = storage.getItem('store_view_currency') || 'USD';
+
+    // Remove attributes not visible on front
+    const cleanAttributes = useMemo(() => {
+        if (customAttributes.length === 0) {
+            return [];
+        }
+
+        return customAttributes
+            .filter(
+                ({ attribute_metadata }) =>
+                    attribute_metadata.is_visible_on_front === true
+            )
+            .sort(
+                (a, b) =>
+                    a.attribute_metadata.sort_order -
+                    b.attribute_metadata.sort_order
+            );
+    }, [customAttributes]);
+
+    const list = useMemo(() => {
+        if (cleanAttributes.length === 0) {
+            return [];
+        }
+
+        return cleanAttributes.map(
```

Comment:
Including the filter and sort iterations above to init `cleanAttributes`, this would be the third pass through these attributes, which could be troublesome as this list of attributes grows. I suggest using a `reduce` call instead, and doing this in a single pass.

```js
customAttributes.reduce( (attributeList, currentAttribute) => {
  // filter
  if (!currentAttribute.attribute_metadata.is_visible_on_front) {
    const attributeContent = existingLogic();
  
    // sort
    attributeList.splice(currentAttribute.attribute_metadata.sort_order, 0, attributeContent)
  }

  return attributeList;
}, [])

Reply:
Thanks for the proposition!
I've moved the sorting logic in the backend to simplify things.